### ISSUE TITLE
Old-style atomics with generic address space

### ISSUE THREAD

**Issue by davidrohr** [KhronosGroup/OpenCL-Docs#66]
Porting some of our OpenCL 1.2 code to OpenCL 2.0, I encountered some problems with atomic operations. After getting rid of the __global and __shared specifiers to pointers, the code fails to compile.

The reason is that the "old-style" atomics we are using, e.g. atomic_add, yield in clang:
address space mismatch in 1st argument ('__generic int *'), parameter type must be 'volatile __global int *'

Unfortunately, switching to the new atomics in C11/C++11 style is not straight forward.
In other parts of the code, where we exclude concurrent access to the integers, we use binary operators (&,  |, ^), which are not available for the C11 atomics like atomic_uint.

Basically, we rely on the fact that the integers behave as normal integers when the pointers are dereferenced, but we need the atomic operations on them in addition (we make sure ourselves, that this will not collide). This is a feature available with the old OpenCL 1.2. atomics, but not with the new C11 atomics.

In my opinion, all operations of "normal" integers should be available also for atomic_uint, etc, or the old-style atomic operations like atomic_add should work with pointers to the generic address space in OpenCL 2.0.

**Comment by presentfactory** [comment]
I am having a similar problem to this due to an algorithm I had written before using the 1.2-style atomics where in most cases non-atomic access would be fine and more performant, but only requiring atomics on the same variables in specific contexts. Additionally due to the fact that there's no specification from what I can see on the size of the new atomic types, I am not even sure how to properly allocate memory for them, for example a buffer of `uint`s one would usually allocate `sizeof(cl_uint) * size` bytes, but if they are `atomic_uint`s, then there's not much that can be done other than hope it matches the size of the non-atomic type (which it should really with how GPUs work usually and given that they're guaranteed to be lockfree).

I'd say there should be some way to use the old behavior or allow for "casting" of the new atomics to their respective types, perhaps as a pointer or something to modify it through non-atomically. Another option could be that since the OpenCL standard for access of non-active union members, perhaps a `uint` unioned with an `atomic_uint` could be defined to work properly depending on which is accessed, but I have no idea really.

**Comment by AnastasiaStulova** [comment]
Also to address the problem with host allocation that @presentfactory mentions would it make sense to allow `_Atomic` specifier? This would mean that all atomic types are represented as regular types with extra atomic specifier. Would this resolve the confusion about allocation of atomic type buffers on the host because it would be just be using the basic type.

`_Atomic int = atomic_int = cl_int`

Although is this the case?

However, relaxing this restriction would mean that frontend needs some way to check allowed atomic types.

**Comment by presentfactory** [comment]
@AnastasiaStulova It might make it a bit more clear, but I still think it causes confusion as looking it at from a C standpoint, nothing about `_Atomic` makes promises about its size due to it being able to be implemented in a manner which has extra data allocated for things such as locks in addition to the value itself. I think all that is needed is simply some sort of mention that the size of atomic types will match their underlying type counterparts, as realistically on the GPU that's pretty much always the case (the only atomics supported are those with hardware support usually such that they don't require locks or other machinery), unless I'm mistaken about that.

**Comment by davidrohr** [comment]
I also do not like it that size of int and atomic int is not necessarily the same. Right now, I am using a static assert in my code to ensure it, but basically I would also prefer if that was ensured by the spec.
I see that this might be different for CPU C compilers, when the processor has no support for atomic operations. But on a GPU, or OpenCL capable device, which is able to process OpenCL old-style atomics, there is no reason to allow the size to be different.

**Comment by AnastasiaStulova** [comment]
I am also trying to understand whether the atomics are intended to be used by both host and device in the system mode, in which case how is the different in types handled?

In case we are going to add cl version of atomic types, would that mean they won't be c11 atomic types any longer?

**Comment by davidrohr** [comment]
@AnastasiaStulova : I can only speak for my case. We basically have two kernels that operate alternatily on a large array of integers.

The first kernel perfoms some logical and integer operations += -= &= |= ^=. Here, it is ensured that each element in the array is accessed by only one thread.

The second kernel performs atomic_max operations with more or less random access.

So, with old-style atomics I just have an array of plain integers, which I use normally in the first kernel, and in the second kernel I use the atomic_max function. Since everything is in global memory, I can basically cast all pointers to (__global volatile int*) when passing them to atomic_max, but that is rather ugly. I'd prefer if the pointer could remain __generic.

When I tried to convert to C99 atomics, the second kernel is not a problem, I can just use atomic_fetch_max, but adapting the first kernel is rather inconvenient.

I would not suggest to add any new cl version of atomic types. Overloading OpenCL 1.2 atomic_max etc. functions for __generic pointers would be fully sufficient. It would not break any existing code, and would make porting of OpenCL 1.2 code to OpenCL 2 much easier. C11 atomics in OpenCL would still behave like on the host.

**Comment by AnastasiaStulova** [comment]
Just to summarize a few action items from this conversation:
1. It seems there is a value in adding overload with generic address space to old style atomics from CL1.2 (i.e. `atomic_max(__generic volatile int*)`) that would be available in CL2.0 mode onward.
2. Supporting operation with CL2.0 atomic types should require frontend only functionality and can be allowed at least as a compiler extension. 
3. There seems to be a need for `cl_*` host atomic types that would have the same bitwidth setup as corresponding types in the kernel code.
4. Not clear of whether we need to provide ability to access CL2.0 atomic type objects both from the host and device...

**Comment by AnastasiaStulova** [comment]
There seems to be no objections to either 1 or 2 and it seems reasonable to add those as Clang extension that can potentially be migrated into Khronos spec.

**Comment by AnastasiaStulova** [comment]
Need conclusion on #65 to proceed.
